src/parser/DASHTree.cpp

```diff
@@ -1021,127 +975,127 @@ void adaptive::CDashTree::ParseTagRepresentation(pugi::xml_node nodeRepr,
+    else if (segTemplate->GetDuration() == 0 && !segTemplate->HasTimeline())
     {
       // In the SegmentTemplate tag must be present the "duration" attribute or the SegmentTimeline tag
       LOG::LogF(LOGWARNING,
                 "Cannot generate segments timeline, SegmentTemplate has no duration attribute.");
     }
     else
     {
-      repr->SetTimescale(segTemplate->GetTimescale());
+      uint64_t segNumber = segTemplate->GetStartNumber();
+      const uint32_t segTimescale = segTemplate->GetTimescale();
+      const uint64_t periodStartMs = period->GetStart() == NO_VALUE ? 0 : period->GetStart();
+      const uint64_t periodStartScaled = periodStartMs * segTemplate->GetTimescale() / 1000;
 
-      // Set the representation duration
-      if (period->GetDuration() > 0)
-      {
-        double durationSecs = static_cast<double>(period->GetDuration()) / period->GetTimescale();
-        repr->SetDuration(static_cast<uint64_t>(durationSecs * segTemplate->GetTimescale()));
-      }
-      else if (adpSet->HasSegmentTimelineDuration())
-      {
-        // Calculate total duration of segments from timeline
-        auto& segTLData = adpSet->SegmentTimelineDuration().GetData();
-        repr->SetDuration(std::accumulate(segTLData.begin(), segTLData.end(), 0ULL));
+      //! @todo: PTO a/v sync to be implemented on session/demuxers
+      const bool hasPTO = segTemplate->HasPresTimeOffset();
 
-      }
-      else if (m_mediaPresDuration > 0)
-      {
-        // Note m_mediaPresDuration include the time of each period, use it may not be always correct
-        // maybe segment timeline generation code should be done after the full manifest parsing
-        // where we can determinate/calculate the duration of each period
-        repr->SetDuration(static_cast<uint64_t>(m_mediaPresDuration * segTemplate->GetTimescale()));
-      }
-      else if (m_timeShiftBufferDepth > 0)
+      if (segTemplate->HasTimeline()) // Generate segments from template timeline
       {
-        repr->SetDuration(static_cast<uint64_t>(m_timeShiftBufferDepth * segTemplate->GetTimescale()));
-      }
-      else // Single segment
-      {
-        repr->SetDuration(segTemplate->GetDuration());
-      }
+        uint64_t time{0};
+        uint64_t totalDuration{0};
 
-      // Calculate total segments
-      size_t segmentsCount{0};
+        for (const auto& tlElem : segTemplate->Timeline())
+        {
+          uint32_t repeat = tlElem.repeat;
+          if (tlElem.time > 0)
+            time = tlElem.time;
 
-      if (adpSet->HasSegmentTimelineDuration())
-      {
-        segmentsCount = adpSet->SegmentTimelineDuration().GetSize();
+          do
+          {
+            CSegment seg;
+            seg.startPTS_ = time;
+            // If no PTO, the "t" value on <SegmentTimeline><S> element should be relative to period start
+            if (!hasPTO)
+              seg.startPTS_ += periodStartScaled;
```

Comment:
If by 'segment pts' you mean the pts of the sample data then this stream is probably trying to simulate having ads inserted which would usually never line up with PTS, compared to streams that just have 'chapters'. I don't know if this helps...

Reply:
> If by 'segment pts' you mean the pts of the

its from manifest, example:
```xml
<Period id="43047706_1" start="PT25610H40M24.076S">
       <AdaptationSet bitstreamSwitching="false" frameRate="25/1" mimeType="video/mp4" segmentAlignment="true" startWithSAP="1" subsegmentAlignment="true" subsegmentStartsWithSAP="1">
            <SegmentTemplate startNumber="1" timescale="90000"/>
            <Representation bandwidth="5000000" codecs="avc1.4d4028" height="1080" id="4" width="1920">
                <SegmentTemplate initialization="asset_1080_5_2init.mp4" media="asset_1080_5_2_$Number%09d$.mp4" startNumber="1" timescale="90000">
                    <SegmentTimeline>
                        <S d="180000" t="0"/>
```
you can see that "t" timestamp of the segment its "0" instead to be within the priod start
usually the "t" timestamp should be as follow:

```xml
<Period duration="PT86.600S" id="43047663" start="PT92198337.476S">
        <AdaptationSet bitstreamSwitching="true" id="792039359" mimeType="video/mp4" segmentAlignment="true" startWithSAP="1" subsegmentAlignment="true" subsegmentStartsWithSAP="1">
            <SegmentTemplate initialization="index_video_$RepresentationID$_0_init.mp4?m=1704732393" media="index_video_$RepresentationID$_0_$Number$.mp4?m=1704732393" presentationTimeOffset="4609916873831" startNumber="43047692" timescale="50000">
                <SegmentTimeline>
                    <S d="100000" r="13" t="4609919673831"/>
```
"t" == 4609919673831
4609919673831 / timescale (50000) = 92198393,476
period start is: 92198337,476
you can see that fall in the period

since in the example above the "t" is "0" segment fall in a bad timing
and so when you check for last segment with `AdaptiveTree::IsLastSegment` is not possible detect the last segment due to wrong segment pts, if you try to remove this code line, you can see that cause a playback problem

but this sample stream play bad also on web player testers, so idk if this is a sample stream bug or something i don't know/understand yet...

Reply:
i do the PR merge and release, we will see how to improve this when nevagive feedback will come